File: Algorithm/Python/115.distinct-subsequences.py
#
# @lc app=leetcode id=115 lang=python
#
# [115] Distinct Subsequences
#

# @lc code=start
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        m, n = len(s), len(t)
        dp = [[0] * (n+1) for _ in range(m+1)]
        for i in range(m+1):
            dp[i][0] = 1

        for i in range(m):
            for j in range(n):
                if i < j:
                    continue
                dp[i+1][j+1] = max(dp[i][j+1], dp[i+1][j+1])
                if s[i] == t[j]:
                    dp[i+1][j+1] = max(dp[i+1][j+1], dp[i][j+1] + dp[i][j])
        return dp[m][n]
        
# @lc code=end

File: Algorithm/C++/115.distinct-subsequences.cpp
/*
 * @lc app=leetcode id=115 lang=cpp
 *
 * [115] Distinct Subsequences
 */

// @lc code=start
#include <string>
#include <vector>

class Solution {
public:
    int numDistinct(std::string s, std::string t) {
        int m = s.size(), n = t.size();
        std::vector<std::vector<uint64_t>> dp(m+1, std::vector<uint64_t>(n+1, 0));
        for (int i = 0; i < m+1; i++) {
            dp[i][0] = 1;
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (j > i) continue;
                dp[i+1][j+1] = std::max(dp[i+1][j+1], dp[i][j+1]);
                if (s[i] == t[j]) {
                    dp[i+1][j+1] = std::max(dp[i+1][j+1], dp[i][j] + dp[i][j+1]);
                }
            }
        }
        return dp[m][n];
    }
};
// @lc code=end

